From e64852e2e40f3657e834dc918c97bb0aac06bf6b Mon Sep 17 00:00:00 2001
From: Arnaud le Taillanter
Date: Thu, 1 Sep 2022 17:34:39 +0200
Subject: [PATCH 4/4] Add unified-split mode to routing config

NB: this change needs to be upstreamed

The values for config_db "docker_routing_config_mode" are:
- separated: FRR config generated by Sonic, each FRR daemon has its own config file
- unified: FRR config generated by Sonic, single FRR config file
- split: FRR config not generated by Sonic, each FRR daemon has its own config file
This commit adds:
- split-unified: FRR config not generated by Sonic, single FRR config file
---
 dockers/docker-fpm-frr/docker_init.sh                      | 3 +++
 dockers/docker-fpm-frr/frr/supervisord/supervisord.conf.j2 | 2 +-
 .../yang-models/sonic-device_metadata.yang                 | 7 ++++++-
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/dockers/docker-fpm-frr/docker_init.sh b/dockers/docker-fpm-frr/docker_init.sh
index 15602f035..1f1bc511d 100755
--- a/dockers/docker-fpm-frr/docker_init.sh
+++ b/dockers/docker-fpm-frr/docker_init.sh
@@ -69,6 +69,9 @@ if [ -z "$CONFIG_TYPE" ] || [ "$CONFIG_TYPE" == "separated" ]; then
 elif [ "$CONFIG_TYPE" == "split" ]; then
     echo "no service integrated-vtysh-config" > /etc/frr/vtysh.conf
     rm -f /etc/frr/frr.conf
+elif [ "$CONFIG_TYPE" == "split-unified" ]; then
+    echo "service integrated-vtysh-config" > /etc/frr/vtysh.conf
+    rm -f /etc/frr/bgpd.conf /etc/frr/zebra.conf /etc/frr/staticd.conf
 elif [ "$CONFIG_TYPE" == "unified" ]; then
     CFGGEN_PARAMS=" \
         -d \
diff --git a/dockers/docker-fpm-frr/frr/supervisord/supervisord.conf.j2 b/dockers/docker-fpm-frr/frr/supervisord/supervisord.conf.j2
index fec72db76..a6c82fd0c 100644
--- a/dockers/docker-fpm-frr/frr/supervisord/supervisord.conf.j2
+++ b/dockers/docker-fpm-frr/frr/supervisord/supervisord.conf.j2
@@ -152,7 +152,7 @@ stderr_logfile=syslog
 dependent_startup=true
 dependent_startup_wait_for=bgpd:running

-{% if DEVICE_METADATA.localhost.docker_routing_config_mode is defined and DEVICE_METADATA.localhost.docker_routing_config_mode == "unified" %}
+{% if DEVICE_METADATA.localhost.docker_routing_config_mode is defined and (DEVICE_METADATA.localhost.docker_routing_config_mode == "unified" or DEVICE_METADATA.localhost.docker_routing_config_mode == "split-unified")%}
 [program:vtysh_b]
 command=/usr/bin/vtysh -b
 priority=6
diff --git a/src/sonic-yang-models/yang-models/sonic-device_metadata.yang b/src/sonic-yang-models/yang-models/sonic-device_metadata.yang
index 172410058..5b06cd631 100644
--- a/src/sonic-yang-models/yang-models/sonic-device_metadata.yang
+++ b/src/sonic-yang-models/yang-models/sonic-device_metadata.yang
@@ -49,8 +49,13 @@ module sonic-device_metadata {
                 }

                 leaf docker_routing_config_mode {
+                    description "This leaf allows different configuration modes for FRR:
+                                - separated: FRR config generated from ConfigDB, each FRR daemon has its own config file
+                                - unified: FRR config generated from ConfigDB, single FRR config file
+                                - split: FRR config not generated from ConfigDB, each FRR daemon has its own config file
+                                - split-unified: FRR config not generated from ConfigDB, single FRR config file";
                     type string {
-                        pattern "unified|split|separated";
+                        pattern "separated|unified|split|split-unified";
                     }
                     default "unified";
                 }
--
2.37.2
